mesa/space.py

```diff
@@ -512,10 +513,11 @@ def position_agent(
         if x == "random" or y == "random":
             if len(self.empties) == 0:
                 raise Exception("ERROR: Grid full")
-            coords = agent.random.choice(sorted(self.empties))
+            num_occupied = self.width * self.height - len(self.empties)
+            self.move_to_empty(agent, num_agents=num_occupied)
```

Comment:
Also, the `num_agents` is an optional argument. When it is `None`, `move_to_empty` automatically computes it via `agent.model.schedule.get_agent_count()`.

Reply:
If `num_agents` is `None`, a test fails.
https://github.com/projectmesa/mesa/blob/88d575d6e33c8b0e992605c043e4b60b0f802366/tests/test_grid.py#L245

```
_________________________________________________________________________ TestSingleGrid.test_enforcement _________________________________________________________________________

self = <mesa.space.SingleGrid object at 0x1290572e0>, agent = <tests.test_grid.MockAgent object at 0x129057a00>, cutoff = 0.998, num_agents = None

    def move_to_empty(
        self, agent: Agent, cutoff: float = 0.998, num_agents: int | None = None
    ) -> None:
        """Moves agent to a random empty cell, vacating agent's old cell."""
        if len(self.empties) == 0:
            raise Exception("ERROR: No empty cells")
        if num_agents is None:
            try:
>               num_agents = agent.model.schedule.get_agent_count()
E               AttributeError: 'MockAgent' object has no attribute 'model'

mesa/space.py:438: AttributeError

During handling of the above exception, another exception occurred:

self = <tests.test_grid.TestSingleGrid testMethod=test_enforcement>

    def test_enforcement(self):
        """
        Test the SingleGrid empty count and enforcement.
        """
    
        assert len(self.grid.empties) == 9
        a = MockAgent(100, None)
        with self.assertRaises(Exception):
            self.grid.place_agent(a, (0, 1))
    
        # Place the agent in an empty cell
>       self.grid.position_agent(a)

tests/test_grid.py:245: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
mesa/space.py:519: in position_agent
    self.move_to_empty(agent)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <mesa.space.SingleGrid object at 0x1290572e0>, agent = <tests.test_grid.MockAgent object at 0x129057a00>, cutoff = 0.998, num_agents = None

    def move_to_empty(
        self, agent: Agent, cutoff: float = 0.998, num_agents: int | None = None
    ) -> None:
        """Moves agent to a random empty cell, vacating agent's old cell."""
        if len(self.empties) == 0:
            raise Exception("ERROR: No empty cells")
        if num_agents is None:
            try:
                num_agents = agent.model.schedule.get_agent_count()
            except AttributeError:
>               raise Exception(
                    "Your agent is not attached to a model, and so Mesa is unable\n"
                    "to figure out the total number of agents you have created.\n"
                    "This number is required in order to calculate the threshold\n"
                    "for using a much faster algorithm to find an empty cell.\n"
                    "In this case, you must specify `num_agents`."
                )
E               Exception: Your agent is not attached to a model, and so Mesa is unable
E               to figure out the total number of agents you have created.
E               This number is required in order to calculate the threshold
E               for using a much faster algorithm to find an empty cell.
E               In this case, you must specify `num_agents`.

mesa/space.py:440: Exception
```

The mock of the agent has no model instance. Is it not intentional?

Reply:
Yeah that wasn't expected. I think we should just use `unittest.mock` to mock the method call.

Reply:
Mocking nested function calls: https://stackoverflow.com/questions/12523189/how-to-mock-nested-functions. Though in our case, it is nested method calls.

Reply:
[66fe552](https://github.com/projectmesa/mesa/pull/1457/commits/66fe5524ef0845f5ffc419ea73b983cdecf6de03) - The test was modified, and `num_occupied` was removed.